ipq806x: Add support for Xiaomi Mi Router HD (R3D)

remittor opened the pull request:
Xiaomi R3D is a 2.4/5 GHz band 11ac router, based on IPQ8064.
----
Specification:
----
* SoC:  Qualcomm IPQ8064
* RAM:  512MB DDR3
* STOR: 256MB NAND (Macronix MX30UF2G18AC-TI)
* ETH:  1x WAN 10/100/1000, 3x LAN 10/100/1000
* WiFi: Qualcomm QCA9984 (5GHz, 4T4R, n/ac)
* WiFi: Qualcomm QCA9980 (2.4GHz, 4T4R, b/g/n)
* USB:  1x 3.0
* SATA: 1x SATA 3.1 (only for internal HDD 3.5")
* BTN:  Power, Reset
* LEDS: Status(Green/Blue/Red)
* UART: present as 4-pads on the PCB (3.3V, 115200-8-N-1)

MAC addresses as verified by stock firmware:
----

| Interface   |       MAC         |  Partition | Offset    | Format |
|-------------|-------------------|------|---------|--------|
| WAN (label) | xx:xx:xx:xx:xx:B2 | ART |0x0     | binary |
| LAN         | xx:xx:xx:xx:xx:B3 | ART | 0x6     | binary |
| WiFi 2g     | xx:xx:xx:xx:xx:B4 | ART | 0x1006  | binary |
| WiFi 5g     | xx:xx:xx:xx:xx:B5 | ART | 0x5006  | binary |

Partition layout and boot:
---
Stock Xiaomi firmware has the MTD split into (among others)

- `kernel0 (@0x0800000)`
- `kernel1 (@0x0c00000)`
- `rootfs0 (@0x1000000)`
- `rootfs1 (@0x3800000)`
- `overlay (@0x6000000)`

Xiaomi uboot expects to find kernels at `0x800000` & `0xc00000` referred to as `system 1` & `system 2` respectively.
a kernel is considered suitable for handing control over if its linux magic number exists & uImage CRC are correct. If either of those conditions fail, a matching sys'n'_fail flag is set in uboot env & a restart performed in the hope that the alternate kernel is okay.
If neither kernel checksums ok and both are marked failed, `system 2` is booted anyway.

Note uboot's `tftp` flash install writes the transferred image to both kernel partitions.

Methods for getting SSH access:
----
1) https://github.com/acecilia/OpenWRTInvasion
2) https://github.com/openwrt-xiaomi/xmir-patcher

Installation:
----
The installation file for OpenWRT is a `*squashfs-factory.bin` file that contains the `kernel` and a `ubi` partition. This is flashed as follows:
```
nvram flag_boot_success=1
nvram flag_boot_rootfs=1
nvram flag_last_success=1
nvram flag_try_sys1_failed=0
nvram flag_try_sys2_failed=0
nvram commit
dd if=factory.bin bs=1M count=4 | mtd write - kernel0
dd if=factory.bin bs=1M count=4 | mtd write - kernel1
dd if=factory.bin bs=1M skip=4  | mtd write - rootfs0
reboot
```
Reverting to stock:
----
Download `facinstall.ipk` from link
https://github.com/openwrt-xiaomi/facinstall/releases/latest and install its.
Install `miwifi_r3d_firmware_XXXXX.bin` image via LuCI interface.

motolav commented:
The kernel module for the fan controller should be in `/package/kernel/linux/modules/hwmon.mk`

remittor commented:
@motolav 
check it: https://github.com/openwrt/openwrt/pull/14332

remittor commented:
@Ansuel , ping

Ansuel commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -56,6 +56,9 @@ case "$FIRMWARE" in
 		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_ascii 0:appsblenv ethaddr) +2)
 		;;
+	xiaomi,mi-router-hd)
+		caldata_extract "ART" 0x1000 0x2f20
```
why no nvmem cell ?

remittor replied:
If these lines are deleted, then Wifi is not initialized:
```
ath10k 6.4 driver, optimized for CT firmware, probing pci device: 0x46.
 ath10k_pci 0000:01:00.0: enabling device (0140 -> 0142)
 ath10k_pci 0000:01:00.0: pci irq msi oper_irq_mode 2 irq_mode 0 reset_mode 0
 ath10k_pci 0000:01:00.0: qca9984/qca9994 hw1.0 target 0x01000000 chip_id 0x00000000 sub 168c:cafe
 ath10k_pci 0000:01:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
 ath10k_pci 0000:01:00.0: firmware ver 10.4b-ct-9984-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 7ea63dc5
 ath10k_pci 0000:01:00.0: failed to fetch board data for bus=pci,vendor=168c,device=0046,subsystem-vendor=168c,subsystem-device=cafe from ath10k/QCA9984/hw1.0/board-2.bin
 ath10k_pci 0000:01:00.0: failed to fetch board-2.bin or board.bin from ath10k/QCA9984/hw1.0
 ath10k_pci 0000:01:00.0: failed to fetch board file: -12
 ath10k_pci 0000:01:00.0: could not probe fw (-12)

 ath10k 6.4 driver, optimized for CT firmware, probing pci device: 0x40.
 ath10k_pci 0001:01:00.0: enabling device (0140 -> 0142)
 ath10k_pci 0001:01:00.0: pci irq msi oper_irq_mode 2 irq_mode 0 reset_mode 0
 ath10k_pci 0001:01:00.0: qca99x0 hw2.0 target 0x01000000 chip_id 0x003b01ff sub 168c:0002
 ath10k_pci 0001:01:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
 ath10k_pci 0001:01:00.0: firmware ver 10.4b-ct-9980-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 b36a12bf
 ath10k_pci 0001:01:00.0: failed to fetch board data for bus=pci,vendor=168c,device=0040,subsystem-vendor=168c,subsystem-device=0002 from ath10k/QCA99X0/hw2.0/board-2.bin
 ath10k_pci 0001:01:00.0: failed to fetch board-2.bin or board.bin from ath10k/QCA99X0/hw2.0
 ath10k_pci 0001:01:00.0: failed to fetch board file: -12
 ath10k_pci 0001:01:00.0: could not probe fw (-12)
```
And I can't understand this strange behavior.

Ansuel replied:
This is strange for one maybe the compatible is wrong. For the other maybe the soc is using a special subsystem kind and we need a special board-2.bin

remittor replied:
When building the firmware, a universal image `board-2.bin` for the Wi-Fi module is placed in directory ` /lib/firmware/$FIRMWARE`.
After the kernel is loaded, the `11-ath10k-caldata` script is launched
```
part=ART
mtd=$(find_mtd_chardev $part)
caldata_dd $mtd /lib/firmware/$FIRMWARE $count $offset
```
This script retrieves the actual `board-2.bin` image from the `ART` partition.
Image `board-2.bin` is placed in directory `/lib/firmware/$FIRMWARE`. And only then Wi-Fi modules are initialized.

I think option `nvmem-cell-names = "pre-calibration"` does something different.

But when using call `caldata_extract "ART"`, everything works fine!

Ansuel replied:
if correctly setup they do the same thing... so it's just that the nvmem-cell is not working... probably the compatible is wrong.

remittor replied:
Fixed

Ansuel commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -92,6 +95,9 @@ case "$FIRMWARE" in
 		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_ascii 0:appsblenv ethaddr) +3)
 		;;
+	xiaomi,mi-router-hd)
+		caldata_extract "ART" 0x5000 0x2f20
```
ditto.

remittor replied:
Fixed

Ansuel commented on target/linux/ipq806x/base-files/etc/init.d/hwmon_fancontrol:
```diff
@@ -0,0 +1,20 @@
+		if [ -n "$path_to_hwmon" ]; then
+			path_to_hwmon=$( dirname "$path_to_hwmon" 2>/dev/null )
+			# Set FAN speed to 80%
+			echo "204" > "$path_to_hwmon/pwm1"
```
This is what is done on stock?

remittor replied:
On the stock firmware, the bash script is immediately launched, which controls the fan speed.
On OpenWRT firmware, the user must install additional software to control the fan. For this router I recommend this: https://github.com/openwrt-xiaomi/openwrt-alpine-fan-control/releases/latest

Ansuel replied:
Yes I mean what is the logic for making the fan turn on or off? Maybe we can describe that in DT with cooling-cell and trip points.

Ansuel replied:
I notice the thing have a temp sensor and the fan... So I guess everything is driven with the temp sensor... I think the 2 can be coupled and by defining some trip point you can describe the FAN function entirely in DT without using userspace scripts.

robimarko replied:
Yes, this should be moved to use the in-kernel thermal trips and fan as the cooling device otherwise we will have a mess with various scripts as each device attempts to use different logic.

remittor replied:
> I think the 2 can be coupled and by defining some trip point you can describe the FAN function entirely in DT without using userspace scripts

It is better to do this as part of a separate pull request, since kernel patches are required.

robimarko replied:
I will ask again:
> Also, how is the EMC230x device even being registered as the current driver is not DT aware at all?

remittor replied:
> Also, how is the EMC230x device even being registered as the current driver is not DT aware at all?

I didn't understand the question

robimarko replied:
> > Also, how is the EMC230x device even being registered as the current driver is not DT aware at all?
> 
> I didn't understand the question

What is registering the EMC230x device as the kernel driver for it does not support device tree at all.

remittor replied:
> Also, how is the EMC230x device even being registered as the current driver is not DT aware at all?

```
root@OpenWrt:~# grep -l -F emc230 /sys/class/hwmon/hwmon*/name
/sys/class/hwmon/hwmon11/name

root@OpenWrt:~# ls -la /sys/class/hwmon/hwmon11
lrwxrwxrwx    1 root     root  0 Sep 10 07:15 /sys/class/hwmon/hwmon11 -> ../../devices/platform/i2c_gpio_0/i2c-0/0-002f/hwmon/hwmon11

root@OpenWrt:~# ls -la /sys/devices/platform/i2c_gpio_0/i2c-0/0-002f/hwmon/hwmon11
drwxr-xr-x    2 root     root             0 Sep 10 07:15 .
drwxr-xr-x    3 root     root             0 Sep 10 07:15 ..
lrwxrwxrwx    1 root     root             0 Sep 10 07:15 device -> ../../../0-002f
-r--r--r--    1 root     root          4096 Sep 12 09:47 fan1_fault
-r--r--r--    1 root     root          4096 Sep 12 09:47 fan1_input
-r--r--r--    1 root     root          4096 Sep 10 07:15 name
lrwxrwxrwx    1 root     root             0 Sep 12 09:47 of_node -> ../../../../../../../firmware/devicetree/base/i2c_gpio_0/fan@2f
-rw-r--r--    1 root     root          4096 Sep 12 09:46 pwm1
lrwxrwxrwx    1 root     root             0 Sep 12 09:47 subsystem -> ../../../../../../../class/hwmon
-rw-r--r--    1 root     root          4096 Sep 12 09:47 uevent

root@OpenWrt:~# cat /sys/devices/platform/i2c_gpio_0/i2c-0/0-002f/hwmon/hwmon11/uevent
OF_NAME=fan
OF_FULLNAME=/i2c_gpio_0/fan@2f
OF_COMPATIBLE_0=microchip,emc2305
OF_COMPATIBLE_N=1
```

robimarko replied:
Do you carry some patches that are not in OpenWrt regarding EMC2305, because the upstream driver does not support DT at all, only manually probing via platform data?

remittor replied:
> Do you carry some patches that are not in OpenWrt regarding EMC2305

There are no patches for this driver.

Ansuel commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-mi-router-hd.dts:
```diff
@@ -0,0 +1,560 @@
+
+	reserved-memory {
+		rsvd@44600000 {
+			reg = <0x44600000 0x200000>; /* IPQ_TZ_APPS_ADDR */
```
mhhh why? don't we already reserve this?

remittor replied:
This was done on the stock firmware. I decided to leave it.

Ansuel replied:
please drop or comment this- We don't use trustzone

remittor replied:
Fixed (drop)

Ansuel commented on target/linux/ipq806x/image/generic.mk:
```diff
@@ -547,6 +547,29 @@ define Device/ubnt_unifi-ac-hd
+	IMAGE/factory.bin := append-kernel | pad-to $$$$(KERNEL_SIZE) | \
+		append-ubi | pad-to $$$$(BLOCKSIZE) | check-size
+	DEVICE_PACKAGES := kmod-i2c-gpio kmod-thermal kmod-hwmon-lm75 \
+		kmod-hwmon-emc2305 hwmon-drivetemp kmod-usb-storage-uas \
```
why usb storage kmod? I think that can be selected later by user? Do we do that also for other devices?

remittor replied:
> why usb storage kmod?
And without this module the USB refuses to work.

> Do we do that also for other devices?
Yes.
cisco_vedge1000, ubnt_edgerouter-e300, wevo_air-duo, dlink_dir-685, dlink_dns-313

robimarko replied:
Generally UAS is not included by default, those that need it can always install it later or via AUC

remittor replied:
> i mean you need block-mount and additional package anyway...
> Generally UAS is not included by default, those that need it can always install it later or via AUC

But the same can be said about `hwmon-drivetemp` driver. 
This driver is pre-installed on many routers.

Also take a look at this place:
https://github.com/openwrt/openwrt/blob/683355d0a6d51cbeda258492bc16a9c6413b387e/target/linux/gemini/image/Makefile#L133-L142

Router Xiaomi R3D is also a NAS system.

remittor replied:
> Generally UAS is not included by default

Fresh commit: https://github.com/openwrt/openwrt/pull/16140
And this commit was checked by @robimarko 

https://github.com/openwrt/openwrt/blob/844701f1c045bf7d9bd06014a4b0cf9dd3d6a29d/target/linux/octeon/image/Makefile#L109

Ansuel commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-mi-router-hd.dts:
```diff
@@ -0,0 +1,560 @@
+		ranges;
+
+		wifi@1,0 {
+			compatible = "pci168c,0040";
```
I think this should be 0046 ?

remittor replied:
You're right! Fixed.

remittor commented:
@Ansuel

Callctack:
```
ath10k_core_probe_fw(ar)
    ath10k_core_pre_cal_download(ar)
        ath10k_download_cal_nvmem(ar, "pre-calibration") 
            devm_nvmem_cell_get(ar->dev, "pre-calibration") 
                nvmem_cell_get(dev, "pre-calibration");
                    of_nvmem_cell_get(dev->of_node, "pre-calibration");
                        np = dev->of_node
                        // of_node.full_name = "/soc/pci@1b500000/bridge@0,0/wifi@1,0"
                        index = of_property_match_string(np, "nvmem-cell-names", id);                        
                        of_parse_phandle_with_optional_args(np, "nvmem-cells", "#nvmem-cell-cells", index, &cell_spec);
                            __of_parse_phandle_with_args(np, "nvmem-cells", "#nvmem-cell-cells", 0, index, &cell_spec);
                                of_for_each_phandle(&it, rc, np, "nvmem-cells", "#nvmem-cell-cells", 0) 
```

## First variant

If the dts contains the following description:
```
&pcie0 {
    bridge@0,0 {
        wifi@1,0 {
            nvmem-cells = <&macaddr_art_1006 0>, <&precal_art_1000>;
            nvmem-cell-names = "mac-address", "pre-calibration";
        };
    };
};    
```
dtb:
```
nand-controller@1ac00000 {
    nand@0 {
        partitions {
            partition@540000 {
                label = "ART";
                nvmem-layout { 
                    macaddr@1006 {
                        reg = <0x1006 0x06>;
                        phandle = <0x2b>;
                    };
                    precal@1000 {
                        reg = <0x1000 0x2f20>;
                        phandle = <0x2c>;
                    };
                };
            };
        };
    };
};
pci@1b500000 {
    bridge@0,0 {
        wifi@1,0 {
            nvmem-cells = <0x2b 0x00 0x2c>;
            nvmem-cell-names = "mac-address\0pre-calibration";
        };
    };
};    
```
Function `__of_parse_phandle_with_args` returns -2
Logs:
```
devm_nvmem_cell_get: id = 'pre-calibration'
of_nvmem_cell_get: id = 'pre-calibration', index = 1
of_nvmem_cell_get: 'bridge@0,0'
of_nvmem_cell_get: 'wifi@1,0'
OF: __of_parse_phandle_with_args: cur_index = 0 , rc = 0 , 2B
OF: __of_parse_phandle_with_args: it.node = 'macaddr@1006'
OF: __of_parse_phandle_with_args: cur_index = 1 , rc = 0 , 0
OF: __of_parse_phandle_with_args: ERR = -2
of_nvmem_cell_get: #nvmem-cell-cells ERROR = -2
```
Instead PHandle = 0x2C loop `of_for_each_phandle` processed 0x00 

## Second variant

If the dts contains the following description:
```
&pcie0 {
    bridge@0,0 {
        wifi@1,0 {
            nvmem-cells = <&macaddr_art_1006>, <&precal_art_1000>;
            nvmem-cell-names = "mac-address", "pre-calibration";
        };
    };
};    
```
dtb:
```
pci@1b500000 {
    bridge@0,0 {
        wifi@1,0 {
            nvmem-cells = <0x2b 0x2c>;
            nvmem-cell-names = "mac-address\0pre-calibration";
        };
    };
};    
```
Function `__of_parse_phandle_with_args` returns -22
Logs:
```
devm_nvmem_cell_get: id = 'pre-calibration'
of_nvmem_cell_get: id = 'pre-calibration', index = 1
of_nvmem_cell_get: 'bridge@0,0'
of_nvmem_cell_get: 'wifi@1,0'
OF: __of_parse_phandle_with_args: cur_index = 0 , rc = 0 , 2B
OF: __of_parse_phandle_with_args: it.node = 'macaddr@1006'
OF: /soc/pci@1b500000/bridge@0,0/wifi@1,0: #nvmem-cell-cells = 1 found 0
OF: __of_parse_phandle_with_args: ERR = -22
of_nvmem_cell_get: #nvmem-cell-cells ERROR = -22
```
Loop `of_for_each_phandle` processed only 0x2B PHandle (macaddr@1006)

## Third variant (worked)

If the dts contains the following description:
```
&pcie0 {
    bridge@0,0 {
        wifi@1,0 {
            nvmem-cells = <&precal_art_1000>, <&macaddr_art_1006>;
            nvmem-cell-names = "pre-calibration", "mac-address";
        };
    };
};    
```
dtb:
```
                    macaddr@1006 {
                        reg = <0x1006 0x06>;
                        phandle = <0x2c>;
                    };
                    precal@1000 {
                        reg = <0x1000 0x2f20>;
                        phandle = <0x2b>;
                    };
pci@1b500000 {
    bridge@0,0 {
        wifi@1,0 {
            nvmem-cells = <0x2b 0x2c>;
            nvmem-cell-names = "pre-calibration\0mac-address";
        };
    };
};    
```
Function `__of_parse_phandle_with_args` returns 0
Logs:
```
devm_nvmem_cell_get: id = 'pre-calibration'
of_nvmem_cell_get: id = 'pre-calibration', index = 1
of_nvmem_cell_get: 'bridge@0,0'
of_nvmem_cell_get: 'wifi@1,0'
OF: __of_parse_phandle_with_args: cur_index = 0 , rc = 0 , 2B
OF: __of_parse_phandle_with_args: it.node = 'precal@1000'
OF: __of_parse_phandle_with_args: FOUND: it.node = 2B 'precal@1000' c = 0
of_nvmem_cell_get: #nvmem-cell-cells: args_count = 0  np = -574900180
ath10k_pci 0000:01:00.0: board_file api 2 bmi_id 0:1 crc32 0bfe5c35
ath10k_pci 0000:01:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
ath10k_pci 0000:01:00.0: msdu-desc: 2500  skid: 32
ath10k_pci 0000:01:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
ath10k_pci 0000:01:00.0: wmi print 'free: 84920 iram: 13156 sram: 11224'
ath10k_pci 0000:01:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-nvmem max-sta 32 raw 0 hwcrypto 1

```
Module `wifi@1,0` successfully initialized!

Ansuel commented:
@remittor mhh can't repro, are you sure you don't have some downstream patch that are messing with this? Anyway whatever doesn't change a thing so i'm ok with it.

Ansuel commented on target/linux/ipq806x/image/generic.mk:
```diff
@@ -547,6 +547,29 @@ define Device/ubnt_unifi-ac-hd
 
+define Device/xiaomi_mi-router-hd
+	$(call Device/LegacyImage)
+	$(Device/dsa-migration)
```
New device no migration?

remittor replied:
My firmware for this router has been used by many users for a long time. My patch was added to Chinese forks a long time ago.
And with kernel version 6.x, DSA support appeared.

robimarko replied:
In that case remove it, forks that supported the device dont matter for this.

remittor replied:
> In that case remove it, forks that supported the device dont matter for this.

Because of this solution, users may encounter problems when installing official OpenWRT firmware.
Why do this?

remittor replied:
Row `$(Device/dsa-migration)` removed

robimarko commented on target/linux/ipq806x/base-files/etc/init.d/bootcount:
```diff
@@ -19,5 +19,13 @@ boot() {
+		local boot_wait=$( fw_printenv boot_wait | cut -d = -f 2 )
+		[ "$boot_wait" != "on" ] && fw_setenv boot_wait on
+		local bootdelay=$( fw_printenv bootdelay | cut -d = -f 2 )
+		[ "$bootdelay" != "3" ] && fw_setenv bootdelay 3
```
0 would mean that you cant interrupt U-Boot at all

Ansuel replied:
Yes so the check should be if it's 0 we set it to a sensible value of 3 seconds... anything else that is not 0 should be 0.

remittor replied:
`bootdelay` fixed

remittor commented on target/linux/ipq806x/base-files/etc/board.d/05_compat-version:
```diff
@@ -32,7 +32,8 @@ case "$(board_name)" in
 	asus,onhub |\
-	tplink,onhub)
+	tplink,onhub|\
+	xiaomi,mi-router-hd)
```
My firmware for this router has been used by many users for a long time. My patch was added to Chinese forks a long time ago.
And with kernel version 6.x, DSA support appeared.

robimarko replied:
In that case remove it, forks that supported the device dont matter for this.

remittor replied:
Removed

Neustradamus commented:
To follow

robimarko commented on package/boot/uboot-envtools/files/ipq806x:
```diff
@@ -56,13 +56,17 @@ qcom,ipq8064-ap148|\
 esac
 
 config_load ubootenv
-config_foreach ubootenv_add_app_config ubootenv
```
Unrelated change

remittor replied:
Let me remind you about this fix: https://github.com/openwrt/openwrt/commit/4cccea02a60aee0dd77c4db35672c92e2fe384a1

Without this patch, the `ubootenv_add_uci_sys_config` settings branch is simply ignored.

robimarko commented on package/boot/uboot-envtools/files/ipq806x:
```diff
@@ -56,13 +56,17 @@ qcom,ipq8064-ap148|\
 	ubootenv_add_uci_config $(ubootenv_mtdinfo)
 	;;
+xiaomi,mi-router-hd)
+	ubootenv_add_uci_config "/dev/mtd9" "0x0" "0x10000" "0x20000"
```
You can use the Linux DT compatible to avoid having to add it here

remittor replied:
All xiaomi routers have two sections that store settings for the bootloader.
It is necessary to allow the user to operate the settings in both sections:
```
	ubootenv_add_uci_config "/dev/mtd9" "0x0" "0x10000" "0x20000"
	ubootenv_add_uci_sys_config "/dev/mtd12" "0x0" "0x10000" "0x20000"
```
Example_1: https://github.com/openwrt/openwrt/blob/9b336455663482739868953b8b1bd46e1dd0943f/package/boot/uboot-envtools/files/qualcommax_ipq807x#L49-L56
Example_2: https://github.com/openwrt/openwrt/blob/9b336455663482739868953b8b1bd46e1dd0943f/package/boot/uboot-envtools/files/mediatek_mt7622#L63-L66

robimarko commented on target/linux/ipq806x/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq8064-mi-router-hd.dts:
```diff
@@ -0,0 +1,555 @@
+	};
+
+	chosen {
+		bootargs = "rootfstype=squashfs noinitrd";
```
I dont think you need `noinitrd`

remittor replied:
Removed from bootargs.

robimarko commented on target/linux/ipq806x/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq8064-mi-router-hd.dts:
```diff
@@ -0,0 +1,555 @@
+		};
+	};
+
+	usb_pwr_en_pins: usb_pwr_en_pins {
```
Better register a regulator that is enabled by default

remittor replied:
Can you show an example?

robimarko replied:
Look at:
https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-rt-ax89x.dts#L144

remittor replied:
Fixed

Neustradamus commented:
@remittor: Any progress on this PR?

remittor commented:
@Neustradamus , I will definitely make corrections within two weeks.

remittor committed:
ipq806x: Add support for Xiaomi Mi Router HD (R3D)

Xiaomi R3D is a 2.4/5 GHz band 11ac router, based on IPQ8064.

Specification:
* SoC:  Qualcomm IPQ8064
* RAM:  512MB DDR3
* STOR: 256MB NAND (Macronix MX30UF2G18AC-TI)
* ETH:  1x WAN 10/100/1000, 3x LAN 10/100/1000
* WiFi: Qualcomm QCA9984 (5GHz, 4T4R, n/ac)
* WiFi: Qualcomm QCA9980 (2.4GHz, 4T4R, b/g/n)
* USB:  1x 3.0
* SATA: 1x SATA 3.1 (only for internal HDD 3.5")
* BTN:  Power, Reset
* LEDS: Status(Green/Blue/Red)
* UART: present as 4-pads on the PCB (3.3V, 115200-8-N-1)

MAC addresses as verified by stock firmware:

| Interface   |       MAC         |  ART    | Format |
|-------------+-------------------+---------+--------|
| WAN (label) | xx:xx:xx:xx:xx:B2 | 0x0     | binary |
| LAN         | xx:xx:xx:xx:xx:B3 | 0x6     | binary |
| WiFi 2g     | xx:xx:xx:xx:xx:B4 | 0x1006  | binary |
| WiFi 5g     | xx:xx:xx:xx:xx:B5 | 0x5006  | binary |

Partition layout and boot:

Stock Xiaomi firmware has the MTD split into (among others)

- kernel0 (@0x0800000)
- kernel1 (@0x0c00000)
- rootfs0 (@0x1000000)
- rootfs1 (@0x3800000)
- overlay (@0x6000000)

Xiaomi uboot expects to find kernels at 0x800000 & 0xc00000
referred to as system 1 & system 2 respectively.
a kernel is considered suitable for handing control over
if its linux magic number exists & uImage CRC are correct.
If either of those conditions fail, a matching sys'n'_fail flag
is set in uboot env & a restart performed in the hope that the
alternate kernel is okay.
If neither kernel checksums ok and both are marked failed, system 2
is booted anyway.

Note uboot's tftp flash install writes the transferred
image to both kernel partitions.

Methods for getting SSH access:
1) https://github.com/acecilia/OpenWRTInvasion
2) https://github.com/openwrt-xiaomi/xmir-patcher

Installation:

The installation file for OpenWRT is a *squashfs-factory.bin file that
contains the kernel and a ubi partition. This is flashed as follows:

nvram flag_boot_success=1
nvram flag_boot_rootfs=1
nvram flag_last_success=1
nvram flag_try_sys1_failed=0
nvram flag_try_sys2_failed=0
nvram commit
dd if=factory.bin bs=1M count=4 | mtd write - kernel0
dd if=factory.bin bs=1M count=4 | mtd write - kernel1
dd if=factory.bin bs=1M skip=4  | mtd write - rootfs0
reboot

Reverting to stock:

Download facinstall.ipk from link
https://github.com/openwrt-xiaomi/facinstall/releases/latest
and install its.
Install miwifi_r3d_firmware_XXXXX.bin image via LuCI interface.

Signed-off-by: Oleg S <remittor@gmail.com>

robimarko commented on target/linux/ipq806x/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq8064-r3d.dts:
```diff
@@ -0,0 +1,562 @@
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x42000000 0x1e000000>;
```
This should get patched in by U-Boot

remittor replied:
So what? State your thoughts in full.

robimarko commented on target/linux/ipq806x/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq8064-r3d.dts:
```diff
@@ -0,0 +1,562 @@
+		scl-gpios = <&qcom_pinmux 54 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+		i2c-gpio,delay-us = <5>;
+
+		fan@2f {
```
None of the DT property you are using for EMC2305 are supported by the mainline driver, it literally does not even support using compatibles

remittor replied:
> DT property you are using for EMC2305

removed

remittor commented on target/linux/ipq806x/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq8064-r3d.dts:
```diff
@@ -0,0 +1,562 @@
+		compatible = "regulator-fixed";
+		regulator-name = "USB_VBUS";
+		//pinctrl-names = "default";
+		//pinctrl-0 = <&pinctrl_usbhost>;        
```
Removed
